module/VuFind/src/VuFind/Search/Memory.php

```diff
@@ -179,4 +228,76 @@ public function retrieveSearch()
     {
         return $this->session->last ?? null;
     }
+
+    /**
+     * Get current search id
+     *
+     * @return ?int
+     */
+    public function getCurrentSearchId(): ?int
+    {
+        return $this->request->getQuery('sid')
+            ?? $this->request->getPost('sid');
+    }
+
+    /**
+     * Get current search
+     *
+     * return ?\VuFind\Search\Base\Results
+     */
+    public function getCurrentSearch(): ?\VuFind\Search\Base\Results
+    {
+        if (!($id = $this->getCurrentSearchId())) {
+            return null;
+        }
+        return $this->getSearchById($id);
+    }
+
+    /**
+     * Get last search id
+     *
+     * @return ?int
+     */
+    public function getLastSearchId(): ?int
+    {
+        $id = $this->request->getQuery('sid')
+            ?? $this->request->getPost('sid')
+            ?? $this->session->lastId;
+        return $id ? (int)$id : null;
+    }
+
+    /**
+     * Get last search
+     *
+     * return ?\VuFind\Search\Base\Results
+     */
+    public function getLastSearch(): ?\VuFind\Search\Base\Results
+    {
+        if (!($id = $this->getLastSearchId())) {
+            return null;
+        }
+        return $this->getSearchById($id);
+    }
+
+    /**
+     * Get a search by id
+     *
+     * @param int $id Search ID
+     *
+     * @return ?\VuFind\Search\Base\Results
+     */
+    protected function getSearchById(int $id): ?\VuFind\Search\Base\Results
+    {
+        if (!array_key_exists($id, $this->searchCache)) {
+            $search
+                = $this->searchTable->getOwnedRowById($id, $this->sessionId, null);
```

Comment:
Don't we also need to pass the user ID here? Or is this the reason you're rewriting the session ID when an existing search is loaded?

Reply:
See above. :)